core/shared/src/main/scala/zio/logging/internal/WriterProvider.scala

```diff
@@ -0,0 +1,83 @@
+package zio.logging.internal
+
+import java.io.{ BufferedWriter, FileOutputStream, OutputStreamWriter, Writer }
+import java.nio.charset.Charset
+import java.nio.file.{ FileSystems, Path }
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import java.time.temporal.ChronoUnit
+
+private[logging] sealed trait WriterProvider {
+  def writer: Writer
+}
+
+private[logging] object WriterProvider {
+  final case class SimpleWriterProvider(
+    destination: Path,
+    charset: Charset,
+    bufferedIOSize: Option[Int]
+  ) extends WriterProvider {
+    override val writer: Writer = {
+      val output = new OutputStreamWriter(new FileOutputStream(destination.toFile, true), charset)
+      bufferedIOSize match {
+        case Some(bufferSize) => new BufferedWriter(output, bufferSize)
+        case None             => output
+      }
+    }
+  }
+
+  final case class TimeBasedRollingWriterProvider(
+    destination: Path,
+    charset: Charset,
+    bufferedIOSize: Option[Int]
+  ) extends WriterProvider {
+    import java.util.concurrent.locks.ReentrantLock
+    import TimeBasedRollingWriterProvider._
+
+    private var timeInUse             = makeNewTime
+    private var currentWriter: Writer = makeWriter(makePath(destination, timeInUse), charset, bufferedIOSize)
+    private val lock: ReentrantLock = new ReentrantLock()
+
+    override def writer: Writer = {
+      val newTime = makeNewTime
+      if (newTime != timeInUse) {
+        lock.lock()
+        try {
+          val newTimeInLock = makeNewTime
+          if (newTimeInLock != timeInUse) {
+            currentWriter.close()
+            currentWriter = makeWriter(makePath(destination, newTimeInLock), charset, bufferedIOSize)
+            timeInUse = newTimeInLock
+          }
+        } finally {
+          lock.unlock()
+        }
+      }
+      currentWriter
+    }
```

Comment:
https://github.com/zio/zio-logging/pull/691#discussion_r1143799344 continuously..

I applied a concurrent pattern and checked the time under the lock.  
After acquiring the lock, I created a new timestamp value using `makeNewTime()`,
. because I think that if I were to use the time that was determined before acquiring the lock to create a new file, repeatedly create and close new writers.

thread
```
A: 
set(newTime) -> lock -> if(newTime..) -> create 
B: 
set(newTime) -> lock -> wait                 -> if(newTime..) -> create
```

Reply:
``` scala 
   override def writer: Writer = {
      val newTime = makeNewTime
      if (newTime != timeInUse) { // if they are same it will not go inside if, so it will not be locked
        lock.lock()
        try {

// considering that time is date, basically it should not happen that newTimeInLock != newTime
// otherwise it will have to wait for lock more almost one day
          val newTimeInLock = makeNewTime 
        
         // considering proposed    if (newTime != timeInUse) this additional check here in lock,
         // this is for situation, when 2 threads at same time try to obtain lock, (after first if condition)
        // one will get lock first, and this will change writer and timeInUse
        // then after lock will be released, second one will get lock, and enter here
        // but timeInUse and writer are already changed, 
        // so it should not be needed to recreate basically same thing
          if (newTimeInLock != timeInUse) {
            currentWriter.close()
            currentWriter = makeWriter(makePath(destination, newTimeInLock), charset, bufferedIOSize)
            timeInUse = newTimeInLock
          }
        } finally {
          lock.unlock()
        }
      }
      currentWriter
    }
```

Reply:
I understand the comments (timeInUse and writer are already changed), and I rewrote code.